From: Fabian Pflug <f.pflug@pengutronix.de>
Date: Wed, 11 Mar 2026 08:59:22 +0100
Subject: [PATCH] common: skov_select_secpolicy: implement choice of security
 policy
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Implement the current best practice and only be less restrictive if the board is
not yet end-use ready or it is fused with the publically known development keys.

Upstream-Status: Pending

Signed-off-by: Fabian Pflug <f.pflug@pengutronix.de>
Signed-off-by: Ulrich Ölmann <u.oelmann@pengutronix.de>
---
 common/Makefile                |  1 +
 common/skov_select_secpolicy.c | 77 ++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 78 insertions(+)
 create mode 100644 common/skov_select_secpolicy.c

diff --git a/common/Makefile b/common/Makefile
index 6b97edc2c334..b6fb11e77f06 100644
--- a/common/Makefile
+++ b/common/Makefile
@@ -80,6 +80,7 @@ obj-$(CONFIG_MENUTREE)		+= menutree.o
 lwl-$(CONFIG_IMD)		+= imd-barebox.o
 obj-$(CONFIG_IMD)		+= imd.o
 obj-y				+= file-list.o
+obj-$(CONFIG_SECURITY_POLICY)	+= skov_select_secpolicy.o
 obj-$(CONFIG_FIRMWARE)		+= firmware.o
 obj-$(CONFIG_UBIFORMAT)		+= ubiformat.o
 obj-$(CONFIG_BAREBOX_UPDATE_IMX_NAND_FCB) += imx-bbu-nand-fcb.o
diff --git a/common/skov_select_secpolicy.c b/common/skov_select_secpolicy.c
new file mode 100644
index 000000000000..696e5f93b8a4
--- /dev/null
+++ b/common/skov_select_secpolicy.c
@@ -0,0 +1,77 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#define pr_fmt(fmt) "skov-secpolicy: " fmt
+
+#include <of.h>
+#include <hab.h>
+#include <init.h>
+#include <linux/string.h> /* for memcmp() */
+#include <linux/printk.h>
+#include <mach/imx/ocotp-fusemap.h>
+#include <security/policy.h>
+
+/*
+ * SRK hash of the publicly available PTX development keys:
+ * https://github.com/pengutronix/ptx-code-signing-dev/tree/master/habv4
+ */
+static u8 srk_hash_devel[] = {
+	0x57, 0x1f, 0x40, 0xf2, 0x8f, 0x8c, 0xdd, 0x35,
+	0xe8, 0x35, 0x70, 0xee, 0x77, 0x23, 0xf0, 0x70,
+	0xd9, 0xa8, 0xbb, 0x76, 0x22, 0x19, 0x2e, 0x7a,
+	0xa2, 0x3f, 0x77, 0xc3, 0x95, 0xeb, 0x92 ,0xaf,
+};
+
+static bool imx_hab_has_devel_keys(void)
+{
+	u8 srk[SRK_HASH_SIZE];
+	int err;
+
+	err = imx_hab_read_srk_hash(srk);
+	if (err) {
+		pr_err("Failed to read SRK hash.\n");
+		return false;
+	}
+
+	/* Compare the burned-in SRK hash against the development key hash. */
+	return memcmp(srk_hash_devel, srk, SRK_HASH_SIZE) == 0;
+}
+
+static int select_security_policy(void)
+{
+	const char *policy = POLICY_LOCKDOWN;
+	bool manufacturing_complete = true;
+	u32 value = U32_MAX;
+	int err = 0;
+
+	/* No support for the outdated i.MX6 and ARM9 boards. */
+	if (!of_machine_is_compatible("skov,imx8mp"))
+		return 0;
+
+	/* Check the manufacturing fuse: OCOTP BANK 14, WORD 0, BIT 0 */
+	err = imx_ocotp_read_field(MX8M_OCOTP_GP1_0, &value);
+	if (err < 0) {
+		pr_err("Failed to read manufacturing fuse.\n");
+		goto out;
+	}
+	manufacturing_complete = value & BIT(0);
+
+	if (!manufacturing_complete) {
+		/* The board is in production. As we lack a factory policy, fall
+		 * back to the devel policy instead */
+		pr_notice("Still in production, switch to devel policy.\n");
+		policy = POLICY_DEVEL;
+	} else if (!imx_hab_has_devel_keys()) {
+		/* The board is end-use ready with secret keys. Choose restrictive policy */
+		pr_notice("End-use ready with secret keys, stay in lockdown policy.\n");
+		policy = POLICY_LOCKDOWN;
+	} else {
+		/* The board is end-use ready with devel keys. Choose permissive policy */
+		pr_notice("End-use ready with devel keys, switch to devel policy.\n");
+		policy = POLICY_DEVEL;
+	}
+
+out:
+	security_policy_select(policy);
+	return 0;
+}
+late_initcall(select_security_policy);
